client/state/data-layer/wpcom/me/settings/index.js

```diff
@@ -23,68 +23,59 @@ import { registerHandlers } from 'state/data-layer/handler-registry';
  * that the REST API returns already HTML-encoded
  */
 const PROPERTIES_TO_DECODE = new Set( [ 'display_name', 'description', 'user_URL' ] );
-function fromApi( apiResponse ) {
-	return mapValues( apiResponse, ( value, name ) => {
-		if ( PROPERTIES_TO_DECODE.has( name ) ) {
-			return decodeEntities( value );
-		}
-
-		return value;
-	} );
-}
+export const fromApi = apiResponse =>
+	mapValues(
+		apiResponse,
+		( value, name ) => ( PROPERTIES_TO_DECODE.has( name ) ? decodeEntities( value ) : value )
+	);
 
 /*
  * Fetch settings from the WordPress.com API at /me/settings endpoint
  */
-export const requestUserSettings = ( { dispatch }, action ) =>
-	dispatch(
-		http(
-			{
-				apiVersion: '1.1',
-				method: 'GET',
-				path: '/me/settings',
-			},
-			action
-		)
+export const requestUserSettings = action =>
+	http(
+		{
+			apiVersion: '1.1',
+			method: 'GET',
+			path: '/me/settings',
+		},
+		action
 	);
 
 /*
  * Store the fetched user settings to Redux state
  */
-export const storeFetchedUserSettings = ( { dispatch }, action, data ) => {
-	dispatch( updateUserSettings( fromApi( data ) ) );
-};
+export const storeFetchedUserSettings = ( action, data ) => updateUserSettings( data );
 
 /*
  * Post settings to WordPress.com API at /me/settings endpoint
  */
-export function saveUserSettings( { dispatch, getState }, action ) {
-	const { settingsOverride } = action;
-	const settings = settingsOverride || getUnsavedUserSettings( getState() );
-
-	if ( ! isEmpty( settings ) ) {
-		dispatch(
-			http(
-				{
-					apiVersion: '1.1',
-					method: 'POST',
-					path: '/me/settings',
-					body: settings,
-				},
-				action
-			)
-		);
-	}
+export function saveUserSettings( action ) {
+	return ( dispatch, getState ) => {
```

Comment:
i'm not sure i've seen returning a thunk from a data layer ex handler before. neat. /cc @jsnajdr

Reply:
This is an unusual case where the payload of the action (`USER_SETTINGS_SAVE`, created by `saveUserSettings()`) doesn't come from action creator parameters, but is read from Redux state. And reading state is possible only in a thunk.

Alternative implementation is to make the `saveUserSettings` action creator thunkish:
```js
const saveUserSettings = settings => ( dispatch, getState ) => {
  settings = settings || getUnsavedUserSettings( getState() );
  dispatch( { type: USER_SETTINGS_SAVE, settings } );
}
```
This action either saves settings explicitly passed as parameter, or, if called without a parameter, reads unsaved settings from Redux state.

If you do this, then the `fetch` handler in data layer becomes trivial -- just return `http()` that uses the `settings` parameter that always has the right value already.

Reply:
Now I realized that _always_ adding the `settings` object to the `USER_SETTINGS_SAVE` would complicate how the `USER_SETTINGS_UNSAVED_CLEAR` works. It uses the `settingsOverride` parameter to distinguish between "save all" and "save just this property" situations and clears the dirty properties accordingly.

Let's keep the patch as it is then. I think it will see some change once we start removing the Flux store for user settings.